File: Examples/Analysis_C/Analysis_2C.m
% Multivariate Bayesian Inversion for Classification and Regression
% Analysis 2: comparison of methods (MATLAB script)
% 
% Author: Joram Soch, OvGU Magdeburg
% E-Mail: joram.soch@ovgu.de
% 
% Version History:
% - 20/12/2024, 17:46: first data analysis
% - 26/02/2025, 16:04: final figure display
% - 26/02/2026, 16:03: rewrote for multiple analyses
% - 26/02/2026, 16:29: added MBC, GNB, LDA, LogReg, SVC, RFC, NNC
% - 26/02/2026, 16:34: restricted to data subsets
% - 26/02/2026, 16:45: remove features for GNB


clear
close all

%%% Step 1: load data %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% load extracted data
load('../Analysis_2/MNIST_data.mat');

% get data dimensions
n1 = size(Y1,1);                % number of training data points
n2 = size(Y2,1);                % number of test data points
v  = size(Y2,2);                % number of features
C  = max(x2);                   % number of classes


%%% Step 2: analyze data %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% specify analysis parameters
meth    = {'MBC', 'GNB', 'LDA', 'LogReg', 'SVC', 'RFC', 'NNC'};
prior.x = [1:C];                % MBC class indices
prior.p = 1/C*ones(1,C);        % MBC prior probabilities
Dgnb    = 'normal';             % GNB distribution name
Dlda    = 'linear';             % LDA discriminant type
Llogreg = 'logit';              % LogReg link function
Mlogreg = 'ordinal';            % LogReg model type
Ilogreg = false;                % LogReg interactions
Csvm    = 1;                    % SVM cost parameter
Mrf     = 'Bag';                % RF aggregation method
Lnn     = [v];                  % NN hidden layer sizes
Ann     = 'sigmoid';            % NN activation function

% restrict to data subset
N1 = 60000;                     % training data points to use
N2 = 10000;                     % test data points to use

% preallocate results
M  = numel(meth);
xp = zeros(N2,M);
CA = zeros( 1,M);

% display parameters
fprintf('\n-> Analysis 2:\n')
fprintf('   - N1 = %d training data points.\n', N1);
fprintf('   - N2 = %d test data points.\n', N2);

% evaluate all methods
for h = 1:M

    % display method
    fprintf('   - %s: ', meth{h});
    
    % Method: multivariate Bayesian classification (MBI)
    % https://github.com/JoramSoch/MBI/blob/main/MATLAB/ML_MBI.m
    if strcmp(meth{h}, 'MBC')
        fprintf('training ... ')
        mba1 = mbitrain(Y1(1:N1,:), x1(1:N1), [], speye(N1), 'MBC');
        fprintf('testing ... ')
        pp2  = mbitest(Y2(1:N2,:), x2(1:N2), [], speye(N2), mba1, prior);
        [m,j]= max(pp2, [], 2);
        xp2  = prior.x(j);
        fprintf('done.\n')
    end;
    
    % Method: Gaussian naive Bayes (MATLAB)
    % https://de.mathworks.com/help/stats/naive-bayes-classification.html
    if strcmp(meth{h}, 'GNB')
        fprintf('remove at-least-one-class zero-variance features ')
        Y1h = Y1(1:N1,:);
        x1h = x1(1:N1);
        Y1v = zeros(C,v);
        for k = 1:C
            Y1v(k,:) = var(Y1h(x1h==k,:));
        end;
        jh  = all(Y1v);
        fprintf('(%d) \n', v-sum(jh));
        fprintf('          ');
        fprintf('training ... ')
        Y1h  = Y1h(:,jh);
        nbc1 = fitcnb(Y1h, x1h, 'DistributionNames', Dgnb, 'Prior', prior.p);
        fprintf('testing ... ')
        Y2h  = Y2(1:N2,jh);
        xp2  = predict(nbc1, Y2h);
        fprintf('done.\n')
    end;
    
    % Method: linear discriminant analysis (MATLAB)
    % https://de.mathworks.com/help/stats/discriminant-analysis.html
    if strcmp(meth{h}, 'LDA')
        fprintf('training ... ')
        lda1 = fitcdiscr(Y1(1:N1,:), x1(1:N1), 'DiscrimType', Dlda);
        fprintf('testing ... ')
        xp2  = predict(lda1, Y2(1:N2,:));
        fprintf('done.\n')
    end;
    
    % Method: multinomial logistic regression (MATLAB)
    % https://de.mathworks.com/help/stats/fitmnr.html
    if strcmp(meth{h}, 'LogReg')
        fprintf('training ... ')
        logreg1 = fitmnr(Y1(1:N1,:), x1(1:N1), 'Link', Llogreg, 'ModelType', Mlogreg, ...
                                               'IncludeClassInteractions', Ilogreg);
        fprintf('testing ... ')
        xp2     = predict(logreg1, Y2(1:N2,:));
        fprintf('done.\n')
    end;
    
    % Method: support vector classification (LibSVM)
    % https://github.com/JoramSoch/ML4ML/blob/main/ML_SVC.m
    if strcmp(meth{h}, 'SVC')
        fprintf('training ... ')
        opt  = sprintf('-s 0 -t 0 -c %s -q', num2str(Csvm));
        svm1 = svmtrain(x1(1:N1), Y1(1:N1,:), opt);
        fprintf('testing ... ')
        xp2  = svmpredict(x2(1:N2), Y2(1:N2,:), svm1, '-q');
        fprintf('done.\n')
    end;

    % Method: random forrest classification (MATLAB)
    % https://de.mathworks.com/help/stats/select-predictors-for-random-forests.html
    if strcmp(meth{h}, 'RFC')
        fprintf('training ... ')
        rfe1 = fitrensemble(Y1(1:N1,:), x1(1:N1), 'Method', Mrf);
        fprintf('testing ... ')
        xp2  = round(predict(rfe1, Y2(1:N2,:)));
        fprintf('done.\n')
    end;

    % Method: neural network classification (MATLAB)
    % https://de.mathworks.com/help/stats/classificationneuralnetwork.html
    if strcmp(meth{h}, 'NNC')
        fprintf('training ... ')
        nn1 = fitcnet(Y1(1:N1,:), x1(1:N1), 'LayerSizes', Lnn, 'Activations', Ann);
        fprintf('testing ... ')
        xp2 = predict(nn1, Y2(1:N2,:));
        fprintf('done.\n')
    end;
    
    % calculate performance
    xp(:,h) = xp2;
    CA(h)   = mean(xp(:,h)==x2(1:N2));

end;
fprintf('\n\n');
clear mba1 nbc1 lda1 logreg1 svm1 rfe1 nn1


%%% Step 3: visualize results %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% open figure
figure('Name', 'Analysis 2 (Comparison)', 'Color', [1 1 1], 'Position', [50 50 800 900]);
cols  = 'bcgmryw';
x_off = 3;
y_off = 0.95;
hold on;

% plot performance
for h = 1:M
    bar(h, CA(h), 0.7, cols(h));
end;
plot([(1-1), (M+1)], [1/C, 1/C], ':k', 'LineWidth', 2);
plot([x_off, x_off]+1/2, [0, 1], '-k', 'LineWidth', 1);
axis([(1-1), (M+1), 0, 1]);
set(gca,'Box','On');
set(gca,'XTick',[1:M],'XTickLabel',meth);
legend([meth, {'chance'}], 'Location', 'Best');
xlabel('classification approach', 'FontSize', 16);
ylabel('classification accuracy', 'FontSize', 16);
title('Analysis 2: Comparison', 'FontSize', 24);
text(x_off+1/2, y_off, sprintf('generative   \nmethods   '), ...
     'FontSize', 16, 'HorizontalAlignment', 'right', 'VerticalAlignment', 'middle');
text(x_off+1/2, y_off, sprintf('   discriminative\n   methods'), ...
     'FontSize', 16, 'HorizontalAlignment', 'left', 'VerticalAlignment', 'middle');

File: Examples/Analysis_C/Analysis_2C.py
# -*- coding: utf-8 -*-
"""
Multivariate Bayesian Inversion for Classification and Regression
Analysis 2: comparison of methods (Python script)

Author: Joram Soch, OvGU Magdeburg
E-Mail: joram.soch@ovgu.de

Version History:
- 2025-01-03, 16:57: first version in MATLAB
- 2025-03-05, 18:44: final version in Python
- 2026-06-19, 12:03: rewrote MBC and SVC,
                     implemented RGA measure
- 2026-06-19, 14:45: added GNB, LDA, LogReg, RFC, NNC
"""


# specify MBI path
MBI_dir = '../../Python/'

# import packages
import os
import numpy as np
import sklearn as skl
import matplotlib.pyplot as plt

# import MBI module
orig_dir = os.getcwd()
os.chdir(MBI_dir)
import MBI
os.chdir(orig_dir)


### Step 1: load data #########################################################

# load extracted data
data = np.load('../Analysis_2/MNIST_data.npz')
Y1   = data['Y1']
Y2   = data['Y2']
x1   = data['x1']
x2   = data['x2']
del data

# get data dimensions
n1   = Y1.shape[0]              # number of training data points
n2   = Y2.shape[0]              # number of test data points
v    = Y2.shape[1]              # number of features
C    = np.max(x2)               # number of classes


### Step 2: analyze data ######################################################

# specify analysis parameters
meth    = ['GNB', 'MBC', 'LDA', 'LogReg', 'SVC', 'RFC', 'NNC']
prior   = {'x': np.arange(C)+1,             # MBC class indices
           'p': 1/C*np.ones(C)}             # MBC prior probabilities
Dgnb    = 'normal'                          # GNB distribution name
Dlda    = 'linear'                          # LDA discriminant type
Llogreg =  1                                # LogReg cost parameter
Ksvm    = 'linear'                          # SVM kernel type
Csvm    =  1                                # SVM cost parameter
Nrf     = 100                               # RF number of trees
Lnn     = (v,)                              # NN hidden layer sizes
Ann     = 'logistic'                        # NN activation function

# restrict to data subset
N1  = 60000                                 # training data points to use
N2  = 10000                                 # test data points to use

# preallocate results
M   = len(meth)
xp  = np.zeros((N2,M))
xs  = np.zeros((N2,C,M))
CA  = np.zeros(M)
RGA = np.zeros(M)

# display parameters
print('\n-> Analysis 2:')
print('   - N1 = {} training data points'.format(N1))
print('   - N2 = {} test data points'.format(N2))

# evaluate all methods
for h in range(M):
    
    # display method
    print('   - {}: '.format(meth[h]), end='')
    
    # Method: multivariate Bayesian classification (MBI)
    # https://github.com/JoramSoch/MBI/blob/main/Python/MBI.py
    if meth[h] == 'MBC':
        print('training ... ', end='')
        mba1 = MBI.model(Y1[:N1,:], x1[:N1], mb_type='MBC').train()
        print('testing ... ', end='')
        pp2  = MBI.model(Y2[:N2,:], x2[:N2], mb_type='MBC').test(mba1, prior)
        xp2  = np.argmax(pp2, axis=1) + 1
        xs2  = pp2
        print('done.')
    
    # # Method: Gaussian naive Bayes (scikit-learn)
    # # https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.GaussianNB.html
    if meth[h] == 'GNB':
        # remove features
        print('remove at-least-one-class zero-variance features ', end='')
        Y1h = Y1[:N1,:]
        x1h = x1[:N1]
        Y1v = np.zeros((C,v))
        for k in range(C):
            Y1v[k,:] = np.var(Y1h[x1h==k+1], axis=0)
        jh  = np.all(Y1v>0, axis=0)
        print('({})'.format(v-np.sum(jh)))
        print('          ', end='')
        # apply classifier
        print('training ... ', end='')
        Y1h  = Y1h[:,jh]
        gnb1 = skl.naive_bayes.GaussianNB().fit(Y1h, x1h)
        print('testing ... ', end='')
        xp2  = gnb1.predict(Y2[:N2,jh])
        xs2  = gnb1.predict_proba(Y2[:N2,jh])
        print('done.')
    
    # Method: linear discriminant analysis (scikit-learn)
    # https://scikit-learn.org/stable/modules/generated/sklearn.discriminant_analysis.LinearDiscriminantAnalysis.html
    if meth[h] == 'LDA':
        print('training ... ', end='')
        lda1 = skl.discriminant_analysis.LinearDiscriminantAnalysis().fit(Y1[:N1,:], x1[:N1])
        print('testing ... ', end='')
        xp2  = lda1.predict(Y2[:N2,:])
        xs2  = lda1.predict_proba(Y2[:N2,:])
        print('done.')
    
    # Method: logistic regression (scikit-learn)
    # https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html
    if meth[h] == 'LogReg':
        print('training ... ')
        logreg1 = skl.linear_model.LogisticRegression(C=Llogreg).fit(Y1[:N1,:], x1[:N1])
        print('testing ... ', end='')
        xp2     = logreg1.predict(Y2[:N2,:])
        xs2     = logreg1.decision_function(Y2[:N2,:])
        print('done.')
    
    # Method: support vector classification (scikit-learn)
    # https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html
    if meth[h] == 'SVC':
        print('training ... ', end='')
        svm1 = skl.svm.SVC(kernel=Ksvm, C=Csvm).fit(Y1[:N1,:], x1[:N1])
        print('testing ... ', end='')
        xp2  = svm1.predict(Y2[:N2,:])
        xs2  = svm1.decision_function(Y2[:N2,:])
        print('done.')
    
    # Method: random forrest classification (scikit-learn)
    # https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html
    if meth[h] == 'RFC':
        print('training ... ', end='')
        rf1 = skl.ensemble.RandomForestClassifier(
            n_estimators=Nrf).fit(Y1[:N1,:], x1[:N1])
        print('testing ... ', end='')
        xp2 = rf1.predict(Y2[:N2,:])
        xs2 = rf1.predict_proba(Y2[:N2,:])
        print('done.')
    
    # Method: neural network classification (scikit-learn)
    # https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html
    if meth[h] == 'NNC':
        print('training ... ', end='')
        nn1 = skl.neural_network.MLPClassifier(
            hidden_layer_sizes=Lnn, activation=Ann).fit(Y1[:N1,:], x1[:N1])
        print('testing ... ', end='')
        xp2 = nn1.predict(Y2[:N2,:])
        xs2 = nn1.predict_proba(Y2[:N2,:])
        print('done.')
    
    # calculate performance
    xp[:,h]   = xp2
    xs[:,:,h] = xs2
    CA[h]     = np.mean(xp[:,h]==x2[:N2])
    RGA[h]    = MBI.calc_RGA(x2[:N2], xs[:,:,h], 'macro')
    
# delete analysis variables  
del mba1, pp2, Y1h, x1h, Y1v, gnb1, lda1, logreg1, svm1, rf1, nn1


### Step 3: visualize results #################################################

# open figure
fig   = plt.figure(figsize=(8,9))
ax    = fig.add_subplot(111)
x_off = 3
y_off = 0.9
cols  = [[  0,  32,  96], [  0,   0, 255], [  0, 176, 240],
         [192,   0,   0], [255,   0,   0], [255, 192,   0], [255, 255,   0]]

# plot performance
for h in range(M):
    col = tuple([rgb/255 for rgb in cols[h]])
    ax.bar(h+1, RGA[h], width=0.7, color=col, edgecolor='k', label=meth[h])
ax.plot([(1-1), (M+1)], [1/2, 1/2],     ':k', linewidth=2, label='chance')
ax.plot([x_off+1/2, x_off+1/2], [0, 1], '-k', linewidth=1)
ax.axis([(1-1), (M+1), 0, 1])
ax.set_xticks([])
ax.legend(loc='lower right', fontsize=12)
ax.set_xlabel('classification approach', fontsize=20)
ax.set_ylabel('rank graduation accuracy', fontsize=20)
ax.set_title('Analysis 2', fontsize=24, fontweight='bold')
ax.tick_params(axis='both', labelsize=12)
ax.text(x_off+1/2, y_off, 'generative   \nmethods   ',
        fontsize=20, ha='right', va='center')
ax.text(x_off+1/2, y_off, '   discriminative\n   methods',
        fontsize=20, ha='left', va='center')

# enable tight layout
fig.tight_layout()